server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,203 @@
+
+        public SnapshotShard(Snapshot snapshot, IndexId index, int shardId) {
+            this.snapshot = snapshot;
+            this.index = index;
+            this.shardId = shardId;
+        }
+
+        public Snapshot getSnapshot() {
+            return snapshot;
+        }
+
+        public IndexId getIndex() {
+            return index;
+        }
+
+        public int getShardId() {
+            return shardId;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final SnapshotShard that = (SnapshotShard) o;
+            return shardId == that.shardId
+                && snapshot.equals(that.snapshot)
+                && index.equals(that.index);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(snapshot, index, shardId);
+        }
+
+        @Override
+        public String toString() {
+            return "SnapshotShard{" +
+                "snapshot=" + snapshot +
+                ", index=" + index +
+                ", shardId=" + shardId +
+                '}';
+        }
+    }
+
+    @Override
+    public SnapshotShardSizeInfo snapshotShardSizes() {
+        return new SnapshotShardSizeInfo(snapshotShardSizes);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
+            final Set<SnapshotShard> requiredSnapshotShards = new HashSet<>();
+            for (ShardRouting shardRouting : event.state().routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+                if (shardRouting.primary() && shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+                    final RecoverySource.SnapshotRecoverySource snapshotRecoverySource =
+                        (RecoverySource.SnapshotRecoverySource) shardRouting.recoverySource();
+                    final SnapshotShard snapshotShard = new SnapshotShard(snapshotRecoverySource.snapshot(),
+                        snapshotRecoverySource.index(), shardRouting.id());
+                    requiredSnapshotShards.add(snapshotShard);
+                    // check if already populated entry
+                    if (snapshotShardSizes.get(snapshotShard) == null) {
+                        // check if already fetching snapshot info in progress
+                        if (fetchingShards.add(snapshotShard)) {
+                            // TODO: Use a dedicated threadpool here? Use FETCH thread pool?
+                            threadPool.generic().execute(new AbstractRunnable() {
+                                @Override
+                                public void onFailure(Exception e) {
+                                    logger.warn(new ParameterizedMessage("failed to retrieve shard size information for {}",
```

Comment:
On failures, every reroute will trigger another round of fetch of shard size(s) and lead to more warning logs. And in the meantime, the cluster is yellow due to `ClusterShardHealth.getInactivePrimaryHealth`.

In particular if this happens together with the bursts described below, it could be a burdensome load on the master node?